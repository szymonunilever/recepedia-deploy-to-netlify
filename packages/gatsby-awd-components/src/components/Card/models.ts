import { ReactChild, ReactElement } from 'react';
import { Icon, RatingAndReviewsProvider, UnileverLibraryComponent } from '../../models';
import { ButtonProps } from '../Button';
import { RatingProps } from '../Rating';

export interface CardProps {
  key: string;
  content: CardContentProps;
  idPropertyName: string;
  ratingWidget?: ReactElement<RatingProps>
  imageSizes: string;
  className?: string;
  brand?: ReactChild;
  children?: ReactElement<ButtonProps> | ReactElement<ButtonProps>[];
}

export interface CardContentProps {
  fields:{
    slug: string
  };
  name?: string;
  title?: string;
  localImage?: Internal.LocalImage;
  images?: Internal.LocalImage[];
  averageRating?: number;
  [key: string]: any;
}

interface CardWrapper {
  key: string;
  children: ReactElement<CardProps>;
  ratingProvider: RatingAndReviewsProvider;
}

export interface RecipeCardWrapperProps extends CardWrapper{}

export interface ProductCardWrapperProps extends CardWrapper{
  mainImageIndex?: number;
}
