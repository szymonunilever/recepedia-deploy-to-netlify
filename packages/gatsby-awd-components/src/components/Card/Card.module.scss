@import '../../scss/config';

.card {
  background-color: $card-bg;
  border-radius: $card-border-radius;
  box-sizing: border-box;
  .card {
    &__top {
      display: flex;
      align-items: center;
    }

    &__content {
      a {
        color: inherit;
        text-decoration: none;
      }
      address {
        font-style: normal;
      }
    }

    &__button {
      @extend %button;
      width: 100%;
      max-width: $card-button-width;
    }
  }
}

/////----------- CUSTOM -----------/////
.card {
  width: 100%;
  padding: 20px 27px;
  margin-bottom: 15px;
  display: flex;
  flex-direction: column;
  @include respondTo(desktop) {
    padding: 29px;
    &:nth-last-child(1) {
      margin-bottom: 0;
    }
  }

  &__icon {
    margin-left: -7px;
    padding-right: 13px;
    @include respondTo(desktop) {
      padding-right: 36px;
    }
    svg {
      width: 22px;
      height: 22px;
      @include respondTo(desktop) {
        width: 32px;
        height: 32px;
      }
    }
  }

  &__title {
    flex: auto;
    font-family: $rubik;
    font-size: 1.6rem;
    line-height: 1.25;
    color: $black-blue;
    margin: 0;
    @include respondTo(desktop) {
      font-size: 2.1rem;
    }
  }

  &__content {
    margin-top: 13.5px;
    line-height: 1.29;
    @include respondTo(desktop) {
      width: 83%;
      margin: 16px auto 0;
    }
    address {
      .card__text {
        margin: 0;
      }
    }
  }

  &__text {
    margin-bottom: 12px;
  }

  &__button {
    @include respondTo(desktop) {
      margin: 70px auto 0;
      max-width: 352px;
    }
  }
}
