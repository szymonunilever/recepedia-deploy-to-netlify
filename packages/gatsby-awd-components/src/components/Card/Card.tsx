import cx from 'classnames';
import React, { FunctionComponent, SyntheticEvent, useContext } from 'react';
import { getImageAlt, iconNormalize } from '../../utils';
import AdaptiveImage from '../AdaptiveImage';
import { ButtonProps } from '../Button';
import { TagName, Text } from '../Text';
import { CardProps } from './models';
import theme from './Card.module.scss';
import { ReactComponent as KnorrLogoIcon } from '../../svgs/inline/logo-knorr.svg';
import { ReactComponent as HellmannsLogoIcon } from '../../svgs/inline/logo-hellmanns-filled.svg';
import { ReactComponent as MaizenaLogoIcon } from '../../svgs/inline/logo-maizena.svg';
import { navigate } from 'gatsby';

export const Card: FunctionComponent<CardProps> = ({
   content,
   idPropertyName,
   children,
   className = '',
   imageSizes,
   ratingWidget,
   brand,
   brandName,
   brandLink,
   showDescription = false,
 }) => {
  const currentBrand = brandName ? brandName.replace(/[^a-zA-Z0-9\s-]+/g, '').toLowerCase() : '';
  const brandsLogo = {
    knorr: KnorrLogoIcon,
    hellmanns: HellmannsLogoIcon,
    maizena: MaizenaLogoIcon,
  };
  const { title, description, fields:{slug}, localImage } = content;
  const itemTitle = title && (
    <Text
      tag={TagName[ `div` ]}
      text={title}
      className={cx(theme.card__title, 'card__title')}
    />
    );
    const descriptionText = (description && showDescription) && (
      <Text
        tag={TagName[ `div` ]}
        text={description}
        className={cx(theme.card__description, 'card__description')}
      />
  );
  const modifiedChildren =
    children &&
    React.Children.map(children, child => {
      return (
        React.isValidElement<ButtonProps>(child) &&
        React.cloneElement<ButtonProps>(child, {
          onClick: (val: boolean) => {
            child &&
              child.props.onClick &&
              child.props.onClick.apply(child.props.onClick, [
                val,
                content[idPropertyName],
              ]);
          },
        })
      );
    });
  const wrapClasses = cx(theme.card, 'card', className);
  const Image = localImage && (
    <AdaptiveImage
      className={cx(theme.card__image, 'card__image')}
      localImage={localImage}
      alt={title ? getImageAlt(title, slug) : 'image'}
      sizes={imageSizes}
    />
  );

  const handleBrandClick = (e: SyntheticEvent) => {
    e.preventDefault();
    e.stopPropagation();
    brandLink && navigate(`${brandLink}?searchQuery=${currentBrand}`);
  };
  return (
    <div className={wrapClasses} data-componentname="card">
      <div className="card__buttons">{modifiedChildren}</div>
      {Image}
      <div className={cx(theme.card__info, 'card__info')}>
        <div className={cx(theme.card__infoText, 'card__info-text')}>
          {itemTitle}
          {brand}
          {ratingWidget}
          {descriptionText}
        </div>
        {(brandName && brandsLogo[currentBrand]) ? (
          <div
            className={cx(theme.card__infoBrand, 'card__info-brand')}
            onClick={handleBrandClick}
          >
            {iconNormalize(brandsLogo[currentBrand])}
          </div>
        ) : null}
      </div>
    </div>
  );
};

export default Card;
