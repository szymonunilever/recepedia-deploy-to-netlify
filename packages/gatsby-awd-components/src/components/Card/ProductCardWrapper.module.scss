@import '../../scss/config';

.product-card{
  position: relative;
  text-decoration: none;
  list-style: none;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-content: center;

  &__title {
    font-size: $card-title-fz;
    font-family: $card-title-ff;
    font-weight: $card-title-fw;
    color: $card-title-color;
    outline: none;
    margin: $card-title-indent;
    @include respondTo(medium) {
      margin: $card-title-indent-lg;
    }
  }

  &__image {
    position: relative;
    min-width: 100%;
    background-color: $white-gray;
  }
  &__buttons {
    position: absolute;
    z-index: 1;
    width: 100%;
  }
}
