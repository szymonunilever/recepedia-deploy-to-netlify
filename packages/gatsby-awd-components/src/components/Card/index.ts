export { default, Card } from './Card';
export * from './models';
