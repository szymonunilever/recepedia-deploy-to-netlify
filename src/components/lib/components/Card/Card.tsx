import cx from 'classnames';
import React from 'react';
import Button from '../Button';
import { TagName, Text } from '../Text';
import theme from './Card.module.scss';
import { CardProps } from './models';
import { navigate } from 'gatsby';
import getComponentDataAttrs from '../../utils/getComponentDataAttrs';

export const Card = ({
  className,
  content,
  content: { title, texts, type, cta },
  Icon,
  titleLevel = 2,
}: CardProps) => {
  const classWrapper = cx(theme.card, className);

  const Title = (
    <Text
      // @ts-ignore
      tag={TagName[`h${titleLevel}`]}
      text={title}
      className="card__title"
    />
  );

  const withType = (
    type: string,
    view: JSX.Element[] | false,
    content: string[] = texts
  ) => {
    if (view) {
      switch (type) {
        case 'phone':
          return <a href={`tel:${content[0]}`}>{view}</a>;
        case 'address':
          return <address>{view}</address>;
        case 'email':
          return <a href={`mailto:${content[0]}`}>{view}</a>;
        default:
          return view;
      }
    }
  };

  const view =
    texts &&
    texts.length > 0 &&
    texts.map((text, key) => (
      <Text key={key} tag={TagName.p} text={text} className="card__text" />
    ));

  const ctaClickHandler = () => {
    if (cta && cta.linkTo) navigate(cta.linkTo);
  };

  return (
    <div {...getComponentDataAttrs('card', content)} className={classWrapper}>
      <div className="card__top">
        {Icon && (
          <div className="card__icon">
            <Icon />
          </div>
        )}
        {Title}
      </div>
      <div className="card__content">
        {withType(type, view)}
        {cta && (
          <Button
            className="card__button"
            content={cta}
            onClick={ctaClickHandler}
          />
        )}
      </div>
    </div>
  );
};

export default Card;
