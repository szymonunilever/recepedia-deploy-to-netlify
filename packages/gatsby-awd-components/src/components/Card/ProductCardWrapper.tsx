import cx from 'classnames';
import React, { FunctionComponent } from 'react';
import { RatingAndReviewsProvider } from '../../models';
import { Rating } from '../Rating';
import { Card } from './Card';
import { ProductCardWrapperProps } from './models';
import theme from './ProductCardWrapper.module.scss';

export const ProductCardWrapper: FunctionComponent<ProductCardWrapperProps> = ({
 children,
 ratingProvider,
 mainImageIndex= 0,
})=>{
  const {averageRating, fields:{slug}, images} = children.props.content;
  const title = children.props.content.title || children.props.content.name;
  const mainImage = images && images[mainImageIndex];
  const localImage = children.props.content.localImage || mainImage;

  // this should be widget for rating for product
  const ratingWidget =
    ratingProvider !== RatingAndReviewsProvider.none ? (
      <></>
    ): undefined;

  return <Card
    {...children.props}
    content={{...children.props.content, title, localImage}}
    ratingWidget={ratingWidget}
    className={theme.productCard}
  />;
};

export default ProductCardWrapper;
