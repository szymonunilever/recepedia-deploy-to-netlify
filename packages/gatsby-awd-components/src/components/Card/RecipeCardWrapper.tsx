import cx from "classnames";
import React, { FunctionComponent } from "react";
import { RatingAndReviewsEntityType, RatingAndReviewsProvider } from "../../models";
import { Rating } from "../Rating";
import { Card } from "./Card";
import theme from "./RecipeCardWrapper.module.scss";
import { RecipeCardWrapperProps } from "./models";

export const RecipeCardWrapper: FunctionComponent<RecipeCardWrapperProps> = ({
children,
ratingProvider,
})=>{
  const {recipeId, averageRating, fields:{slug}} = children.props.content;
  const ratingWidget =
    ratingProvider !== RatingAndReviewsProvider.none ? (
        <Rating
          className={cx(theme.recipeRatingStars, 'recipe-rating--stars')}
          id={recipeId}
          provider={ratingProvider}
          averageRating={averageRating}
          linkTo={slug}
          entityType={RatingAndReviewsEntityType.recipe}
        />
    ): undefined;

  return <Card {...children.props} ratingWidget={ratingWidget} className={theme.recipeCard} />;
};

export default RecipeCardWrapper;
