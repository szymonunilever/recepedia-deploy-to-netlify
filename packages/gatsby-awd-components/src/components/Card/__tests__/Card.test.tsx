import { mount, ReactWrapper } from 'enzyme';
import React from 'react';
import { Card } from '../Card';
import toJson from 'enzyme-to-json';
import cardsData from 'src/mocks/cards.json';

describe('<Card />', () => {
  let wrapper: ReactWrapper;
  const cards: any = cardsData;

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches the snapshot', () => {
    wrapper = mount(<Card {...cards[0]} Icon="svg" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should show address card', () => {
    wrapper = mount(<Card {...cards[1]} Icon="svg" />);
    expect(wrapper.find('address')).toHaveLength(1);
  });

  it('should show text card', () => {
    wrapper = mount(<Card {...cards[2]} Icon="svg" />);
    expect(wrapper.find('.card__text')).toHaveLength(2);
  });
});
