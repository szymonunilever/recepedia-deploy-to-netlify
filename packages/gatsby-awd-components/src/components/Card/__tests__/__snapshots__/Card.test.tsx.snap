// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> matches the snapshot 1`] = `
<Component
  Icon="svg"
  content={
    Object {
      "cta": Object {
        "label": "Contact Form",
        "linkTo": "/contact",
      },
      "texts": Array [
        "+380633334455",
      ],
      "title": "Phone",
      "type": "phone",
    }
  }
  titleLevel={2}
>
  <div
    className="card"
    data-componentname="card"
    data-componentview="card"
  >
    <div
      className="card__top card__top"
    >
      <div
        className="card__icon card__icon"
      >
        <svg />
      </div>
      <Text
        className="card__title card__title"
        tag={1}
        text="Phone"
      >
        <h2
          className="card__title card__title"
        >
          Phone
        </h2>
      </Text>
    </div>
    <div
      className="card__content card__content"
    >
      <a
        href="tel:+380633334455"
      >
        <Text
          className="card__text card__text"
          key="0"
          tag={7}
          text="+380633334455"
        >
          <p
            className="card__text card__text"
          >
            +380633334455
          </p>
        </Text>
      </a>
      <Component
        className="card__button card__button"
        content={
          Object {
            "label": "Contact Form",
            "linkTo": "/contact",
          }
        }
        onClick={[Function]}
      >
        <button
          className="button button card__button card__button"
          data-componentname="button"
          data-componentview="button"
          disabled={false}
          hidden={false}
          onClick={[Function]}
          type="button"
        >
          <span
            className="button__label"
          >
            Contact Form
          </span>
        </button>
      </Component>
    </div>
  </div>
</Component>
`;
